check: fix compilation on cross-compiled musl

Kiskae opened the pull request:
for some reason it detects static musl as having a non-compliant vsnprintf implementation.

It then tries to provide its own implementation, but it conflicts with `FORTIFY_SOURCE`.

`FORTIFY_SOURCE`  error:
```console
       > In file included from /nix/store/0q2slv31qy84nlqqbhv0bfzxbf5ryqhg-fortify-headers-1.1alpine3/include/stdio.h:26,
       >                  from libcompat.h:99,
       >                  from libcompat.c:21:
       > /nix/store/0q2slv31qy84nlqqbhv0bfzxbf5ryqhg-fortify-headers-1.1alpine3/include/stdio.h:80:1: error: 'vsnprintf' undeclared here (not in a function); did you mean 'vsprintf'?
       >    80 | _FORTIFY_FN(vsnprintf) int vsnprintf(char * _FORTIFY_POS0 __s, size_t __n,
       >       | ^~~~~~~~~~~
       > compilation terminated due to -Wfatal-errors.
```

Detection of non-compliant implementation:
```console
check> checking whether vsnprintf is C99 compliant... no
check> checking whether snprintf is C99 compliant... no
```

## Description of changes

I've disabled the autoconf check for compliance for static stdenv, but I'm not sure how to check whether the implementation is actually compliant or not.

Since I've edited the autoconf config, I've also added the necessary dependencies to rebuild the makefiles.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Kiskae commented:
Rebased to `staging` since it appears to affect a lot of downstream packages and its not that urgent.

yu-re-ka commented:
`hardeningDisable = [ "fortify" ];` would probably also be fine

Kiskae commented:
> `hardeningDisable = [ "fortify" ];` would probably also be fine

Yup, that seems less hacky as well. I'm still not sure why specifically the static stdenv is detected as non-compliant with the c99 standard.

yu-re-ka commented:
It's because of how the fortify hardening is implemented in the fortify-headers for musl.

Kiskae commented:
It doesn't appear to be related to musl, building `pkgMusl.check` shows that it is detected as compliant:

```console
❯ nom build .#pkgsMusl.check -L --keep-going
check> checking whether vsnprintf is C99 compliant... yes
check> checking whether snprintf is C99 compliant... yes
```

yu-re-ka commented:
I had this with x86_64-unknown-linux-gnu -> aarch64-unknown-linux-musl earlier

Kiskae commented:
> I had this with x86_64-unknown-linux-gnu -> aarch64-unknown-linux-musl earlier, so maybe the issue only occurs when cross-compiling to musl

After testing `pkgsCross.aarch64-multiplatform` and `pkgsCross.aarch64-multiplatform-musl`, I can confirm this appears to be triggered by cross-compiling to musl. I'll adjust the check accordingly.

yu-re-ka reviewed (commented):
the hardeningDisable change looks fine.  I can not judge the 'Test can randomly fail'  though. Hydra logs are missing.

Kiskae commented:
That was just a statement that got moved down to group the build attributes together. Although it appears to have been made 10 years ago so the tests might work now: https://github.com/NixOS/nixpkgs/commit/21f4b5c2f7a6b345a0b57eab61f08600a5acbcf0

yu-re-ka commented on pkgs/development/libraries/check/default.nix:
```diff
@@ -11,11 +15,16 @@ stdenv.mkDerivation rec {
+  # fortify breaks the libcompat vsnprintf implementation
+  hardeningDisable = lib.optionals (stdenv.hostPlatform.isMusl && (stdenv.hostPlatform != stdenv.buildPlatform)) [ "fortify" ];
+
+  nativeBuildInputs = [ autoreconfHook pkg-config ];
```
why the autoreconfHook? pkgsStatic.check also builds without it

Kiskae replied:
Ah, that was required when the fix involved changing the configure.ac file since it invalidated the `configure` script included with the release.

I can remove it unless there is an upside to generating the autoconf script within the nix sandbox?

yu-re-ka replied:
Yeah, please remove it

yu-re-ka commented:
Yeah, please remove it

Kiskae committed:
check: fix compilation on cross-compiled musl

on cross-compiled musl vsnprintf gets detected as non-compliant with c99.
The included replacement is incompatible with SOURCE_FORTIFY.

yu-re-ka commented:
Now this shouldn't be a mass-rebuild anymore (as ofborg will tell us in a bit). Can you rebase to master as described in CONTRIBUTING.md?

Kiskae commented:
It still adds `"hardeningDisable": ""` to the derivation for native builds, so I think it will still cause a mass-rebuild. If ofborg says otherwise I'll rebase.

yu-re-ka commented:
Ah, no, you're right
